### ISSUE TITLE
add-config resets all topic configuration to default values

### ISSUE THREAD

**Issue by la-mar** [birdayz/kaf#188]
# Summary
There appears to be a bug in the `kaf topic add-config` command: subsequent calls to `kaf topic add-config` that set different configuration keys cause all other configuration keys to revert to their default values.
I am not entirely clear on if this is a bug or the intended behavior.  The implementation for both `add-config` and `set-config` look straightforward, the only apparent difference being that `set-config` allows you to set multiple k/v pairs at once via a comma-delimited string.  Both commands ultimately pass the config values they were given to [sarama.ClusterAdmin.AlterConfig](https://github.dev/Shopify/sarama/blob/bad67e5b089437bc73f8034a95017e77be71e8b0/admin.go#L66-L71). The docstring on `sarama.ClusterAdmin.AlterConfig` mentions updating the topic with default values, it isn't clear what that means exactly or if it is relevant to this issue.

Steps to reproduce

1. Create topic:
```sh
> toolbox ~ $ kaf topic create conftest -p 10
✅ Created topic!
      Topic Name:            conftest
      Partitions:            10
      Replication Factor:    1
      Cleanup Policy:        delete
```

2. View default topic config
```sh
> toolbox ~ $ kaf topic describe conftest
Name:        conftest
Internal:    false
Compacted:   false
Partitions:
  Partition  High Watermark  Leader  Replicas  ISR
  ---------  --------------  ------  --------  ---
  0          0               2       [1 2 3]   [1 2 3]
  1          0               1       [1 2 3]   [1 2 3]
  2          0               3       [1 2 3]   [1 2 3]
  3          0               2       [1 2 3]   [1 2 3]
  4          0               1       [1 2 3]   [1 2 3]
  5          0               3       [1 2 3]   [1 2 3]
  6          0               2       [1 2 3]   [1 2 3]
  7          0               1       [1 2 3]   [1 2 3]
  8          0               3       [1 2 3]   [1 2 3]
  9          0               2       [1 2 3]   [1 2 3]
Summed HighWatermark:  0
Config:
  Name                            Value      ReadOnly  Sensitive
  ----                            -----      --------  ---------
  min.insync.replicas             2          false     false        # note ISR is 2
  cleanup.policy                  delete     false     false
  message.format.version          2.2-IV1    false     false
  unclean.leader.election.enable  true       false     false
  segment.ms                      604800000  false     false
```

3. Add a configuration to the topic via `add-config`
```sh
> toolbox ~ $ kaf topic add-config conftest min.insync.replicas 1
Added config min.insync.replicas=1 to topic conftest.
```

4. Confirm config was updated
```sh
> toolbox ~ $ kaf topic describe conftest
Name:        conftest
Internal:    false
Compacted:   false
Partitions:
  Partition  High Watermark  Leader  Replicas  ISR
  ---------  --------------  ------  --------  ---
  0          0               2       [1 2 3]   [1 2 3]
  1          0               1       [1 2 3]   [1 2 3]
  2          0               3       [1 2 3]   [1 2 3]
  3          0               2       [1 2 3]   [1 2 3]
  4          0               1       [1 2 3]   [1 2 3]
  5          0               3       [1 2 3]   [1 2 3]
  6          0               2       [1 2 3]   [1 2 3]
  7          0               1       [1 2 3]   [1 2 3]
  8          0               3       [1 2 3]   [1 2 3]
  9          0               2       [1 2 3]   [1 2 3]
Summed HighWatermark:  0
Config:
  Name                            Value      ReadOnly  Sensitive
  ----                            -----      --------  ---------
  min.insync.replicas             1          false     false        # note ISR changed to 1
  message.format.version          2.2-IV1    false     false
  unclean.leader.election.enable  true       false     false
  segment.ms                      604800000  false     false
```

5. Add another configuration to the topic via `add-config`
```sh
> toolbox ~ $ kaf topic add-config conftest retention.ms 604800000
Added config retention.ms=604800000 to topic conftest.
```

6. Viewing the topic config again shows that ISR was reverted to 2 (its original value) and retention.ms was successfully added.
```sh
> toolbox ~ $ kaf topic describe conftest
Name:        conftest
Internal:    false
Compacted:   false
Partitions:
  Partition  High Watermark  Leader  Replicas  ISR
  ---------  --------------  ------  --------  ---
  0          0               2       [1 2 3]   [1 2 3]
  1          0               1       [1 2 3]   [1 2 3]
  2          0               3       [1 2 3]   [1 2 3]
  3          0               2       [1 2 3]   [1 2 3]
  4          0               1       [1 2 3]   [1 2 3]
  5          0               3       [1 2 3]   [1 2 3]
  6          0               2       [1 2 3]   [1 2 3]
  7          0               1       [1 2 3]   [1 2 3]
  8          0               3       [1 2 3]   [1 2 3]
  9          0               2       [1 2 3]   [1 2 3]
Summed HighWatermark:  0
Config:
  Name                            Value      ReadOnly  Sensitive
  ----                            -----      --------  ---------
  min.insync.replicas             2          false     false
  retention.ms                    604800000  false     false
  message.format.version          2.2-IV1    false     false
  unclean.leader.election.enable  true       false     false
  segment.ms                      604800000  false     false
  ```

  Thanks in advance for any commentary you can provide on this issue.

**Comment by birdayz** [comment]
This is indeed interesting..i haven't found a flag or an explicit call to avoid this behavior. a workaround could be, reading the current config first, and setting all entries. should not be too difficult..but it's not transactional. i wonder, if there is a "proper" way? at least in sarama's config i don't see anything else. There is `IncrementalAlterConfig` but it's also setting defaults.

**Comment by la-mar** [comment]
I noticed the lack of alternatives in the sarama codebase as well.  Interestingly, using Kafka's builtin shell script appears to work as expected:

# Summary
There appears to be a bug in the `kaf topic add-config` command: subsequent calls to `kaf topic add-config` that set different configuration keys cause all other configuration keys to revert to their default values.
I am not entirely clear on if this is a bug or the intended behavior.  The implementation for both `add-config` and `set-config` look straightforward, the only apparent difference being that `set-config` allows you to set multiple k/v pairs at once via a comma-delimited string.  Both commands ultimately pass the config values they were given to [sarama.ClusterAdmin.AlterConfig](https://github.dev/Shopify/sarama/blob/bad67e5b089437bc73f8034a95017e77be71e8b0/admin.go#L66-L71). The docstring on `sarama.ClusterAdmin.AlterConfig` mentions updating the topic with default values, it isn't clear what that means exactly or if it is relevant to this issue.


Steps to reproduce

1. Create topic:
```sh
> toolbox ~ $ kaf topic create conftest -p 10
✅ Created topic!
      Topic Name:            conftest
      Partitions:            10
      Replication Factor:    1
      Cleanup Policy:        delete
```

2. View default topic config
```sh
> toolbox ~ $ kaf topic describe conftest
Name:        conftest
Internal:    false
Compacted:   false
Partitions:
  Partition  High Watermark  Leader  Replicas  ISR
  ---------  --------------  ------  --------  ---
  0          0               2       [1 2 3]   [1 2 3]
  1          0               1       [1 2 3]   [1 2 3]
  2          0               3       [1 2 3]   [1 2 3]
  3          0               2       [1 2 3]   [1 2 3]
  4          0               1       [1 2 3]   [1 2 3]
  5          0               3       [1 2 3]   [1 2 3]
  6          0               2       [1 2 3]   [1 2 3]
  7          0               1       [1 2 3]   [1 2 3]
  8          0               3       [1 2 3]   [1 2 3]
  9          0               2       [1 2 3]   [1 2 3]
Summed HighWatermark:  0
Config:
  Name                            Value      ReadOnly  Sensitive
  ----                            -----      --------  ---------
  min.insync.replicas             2          false     false        # note ISR is 2
  cleanup.policy                  delete     false     false
  message.format.version          2.2-IV1    false     false
  unclean.leader.election.enable  true       false     false
  segment.ms                      604800000  false     false
```

3. Add a configuration to the topic via `add-config`
```sh
> toolbox ~ $ kaf topic add-config conftest min.insync.replicas 1
Added config min.insync.replicas=1 to topic conftest.
```

4. Confirm config was updated
```sh
> toolbox ~ $ kaf topic describe conftest
Name:        conftest
Internal:    false
Compacted:   false
Partitions:
  Partition  High Watermark  Leader  Replicas  ISR
  ---------  --------------  ------  --------  ---
  0          0               2       [1 2 3]   [1 2 3]
  1          0               1       [1 2 3]   [1 2 3]
  2          0               3       [1 2 3]   [1 2 3]
  3          0               2       [1 2 3]   [1 2 3]
  4          0               1       [1 2 3]   [1 2 3]
  5          0               3       [1 2 3]   [1 2 3]
  6          0               2       [1 2 3]   [1 2 3]
  7          0               1       [1 2 3]   [1 2 3]
  8          0               3       [1 2 3]   [1 2 3]
  9          0               2       [1 2 3]   [1 2 3]
Summed HighWatermark:  0
Config:
  Name                            Value      ReadOnly  Sensitive
  ----                            -----      --------  ---------
  min.insync.replicas             1          false     false        # note ISR changed to 1
  message.format.version          2.2-IV1    false     false
  unclean.leader.election.enable  true       false     false
  segment.ms                      604800000  false     false
```

5. Add another configuration to the topic via `kafka-configs.sh`:
```sh
> toolbox ~ $ kafka-configs.sh --zookeeper $ZOOKEEPER --entity-type topics --entity-name $TOPIC_NAME --alter --add-config retention.ms=604800000
Completed Updating config for entity: topic 'conftest'.
```

6. Viewing the topic config again shows that ISR override remained in place and retention.ms was successfully added.
```sh
> toolbox ~ $ kaf topic describe conftest
Name:        conftest
Internal:    false
Compacted:   false
Partitions:
  Partition  High Watermark  Leader  Replicas  ISR
  ---------  --------------  ------  --------  ---
  0          0               3       [3]       [3]
  1          0               2       [2]       [2]
  2          0               1       [1]       [1]
  3          0               3       [3]       [3]
  4          0               2       [2]       [2]
  5          0               1       [1]       [1]
  6          0               3       [3]       [3]
  7          0               2       [2]       [2]
  8          0               1       [1]       [1]
  9          0               3       [3]       [3]
Summed HighWatermark:  0
Config:
  Name                            Value      ReadOnly  Sensitive
  ----                            -----      --------  ---------
  min.insync.replicas             1          false     false
  retention.ms                    604800000  false     false
  message.format.version          2.2-IV1    false     false
  unclean.leader.election.enable  true       false     false
  segment.ms                      604800000  false     false
  ```

Perhaps it is a deviation in Sarama's implementation of the Kafka Protocol, though what it may be doesn't jump out to me when poking around their [AlterConfig](https://github.dev/Shopify/sarama/blob/bad67e5b089437bc73f8034a95017e77be71e8b0/admin.go#L680-L730) implementation.  I didn't find anything in [librdkafka](https://github.dev/edenhill/librdkafka/blob/75b60708e6b3bb4984a090efded0d5d209cb7be0/src/rdkafka_admin.c#L2557-L2565) or a similar implementation in Rust ([rust-rdkafka](https://github.dev/fede1024/rust-rdkafka/blob/221fb0540c4cee3c1805bac16f3ab44282263fe7/src/admin.rs#L1081-L1132)) that uses librdkafka directly to suggest there is a need or option to first read the current config, merge it with the updated config items, then reapply.  

It's a bit of a puzzler.

**Comment by birdayz** [comment]
Fixed in master and v0.2.3. This was expected behavior. In the fix, we use `IncrementalAlterConfig` which has the desired "patch" semantics.
